wasmsdk/player_stream.go

```diff
@@ -0,0 +1,230 @@
+	"github.com/0chain/gosdk/zboxcore/sdk"
+)
+
+type StreamPlayer struct {
+	allocationID string
+	remotePath   string
+	authTicket   string
+	lookupHash   string
+
+	isViewer      bool
+	allocationObj *sdk.Allocation
+	authTicketObj *marker.AuthTicket
+
+	waitingToDownloadFiles      chan sdk.PlaylistFile
+	latestWaitingToDownloadFile sdk.PlaylistFile
+
+	downloadedFiles chan []byte
+	ctx             context.Context
+	cancel          context.CancelFunc
+	prefetchQty     int
+
+	timer jsbridge.Timer
+}
+
+func (p *StreamPlayer) Start() error {
+	if p.cancel != nil {
+		p.cancel()
+	}
+
+	p.ctx, p.cancel = context.WithCancel(context.TODO())
+	p.waitingToDownloadFiles = make(chan sdk.PlaylistFile, p.prefetchQty)
+	p.downloadedFiles = make(chan []byte, p.prefetchQty)
+
+	p.timer = *jsbridge.NewTimer(5*time.Second, p.reloadList)
+	p.timer.Start()
+
+	go p.reloadList()
+	go p.startDownload()
+
+	return nil
+}
+
+func (p *StreamPlayer) Stop() {
+	if p.cancel != nil {
+		p.timer.Stop()
+		p.cancel()
+		p.cancel = nil
+	}
+
+	close(p.downloadedFiles)
+}
+
+func (p *StreamPlayer) download(it sdk.PlaylistFile) {
+	wg := &sync.WaitGroup{}
+	statusBar := &StatusBar{wg: wg}
+	wg.Add(1)
+
+	fileName := it.Name
+	localPath := filepath.Join(p.allocationID, fileName)
+
+	downloader, err := sdk.CreateDownloader(p.allocationID, localPath, it.Path,
+		sdk.WithAllocation(p.allocationObj),
+		sdk.WithAuthticket(p.authTicket, p.lookupHash))
+
+	if err != nil {
+		PrintError(err.Error())
+		return
+	}
+
+	defer sys.Files.Remove(localPath) //nolint
+
+	PrintInfo("playlist: downloading [", it.Path, "]")
+	err = downloader.Start(statusBar)
+
+	if err == nil {
+		wg.Wait()
+	} else {
+		PrintError("playlist: download failed.", err.Error())
+		return
+	}
+	if !statusBar.success {
+		PrintError("playlist: download failed: unknown error")
+		return
+	}
+
+	PrintInfo("playlist: downloaded [", it.Path, "]")
+	fs, _ := sys.Files.Open(localPath)
+
+	mf, _ := fs.(*sys.MemFile)
+
+	withRecover(func() {
+		if p.downloadedFiles != nil {
+			p.downloadedFiles <- mf.Buffer.Bytes()
+		}
+	})
+}
+
+func (p *StreamPlayer) startDownload() {
+	for {
+		select {
+		case <-p.ctx.Done():
+			PrintInfo("playlist: download is cancelled")
+			close(p.waitingToDownloadFiles)
+			return
+		case it, ok := <-p.waitingToDownloadFiles:
+			if ok {
+				p.download(it)
+			}
+		}
+	}
+}
+
+func (p *StreamPlayer) reloadList() {
+
+	// `waiting to download files` buffer is too less, try to load latest list from remote
+	if len(p.waitingToDownloadFiles) < p.prefetchQty {
+
+		list, err := p.loadList()
+
+		if err != nil {
+			PrintError(err.Error())
+			return
+		}
+
+		PrintInfo("playlist: ", len(list))
+
+		for _, it := range list {
+			PrintInfo("playlist: +", it.Path)
+
+			if !withRecover(func() {
+				if p.waitingToDownloadFiles != nil {
+					p.waitingToDownloadFiles <- it
+				}
+			}) {
+				// player is stopped
+				return
+			}
+
+			p.latestWaitingToDownloadFile = it
```

Comment:
Seems this is not thread safe, as the reloadList would be called concurrently in different thread. Please make sure it's safe in all the places where it is used.

Reply:
fixed